CONSOLE-2425: Support localization of dynamic plugins

jhadvig opened the pull request:
Implementing additional endpoint for handling locales. The endpoint has following format:
`api/locales/<language>/<plugin-locale-file>`
eg. `api/locales/en/olm.json`

When the request hits the endpoint:
1. it will parse out language and plugin's locale file
2. then it will compare the plugin name against the list of enabled plugins
3. if it matches any of the enabled plugins, it will redirect to its service
4. if it doesn't match any of the enabled plugins, use redirect to `loadPath`, which is set to `static/locales/{{lng}}/{{ns}}.json`

Hopefully I understood the desired logic correctly.
@spadgett shall I also update the i18n enhancement ?

PTAL

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -26,6 +26,79 @@ func NewPluginsHandler(client *http.Client, token string, pluginsEndpointMap map
+
+// parseLangAndPluginName will parse the language and plugin locale file
+func parseLangAndPluginLocaleFile(urlPath string) (string, string, error) {
+	langAndPluginLocaleFile := strings.SplitN(urlPath, "/", 2)
```
I wish we were using something like https://github.com/gorilla/mux to handle variables in the path, although that seems like overkill for one endpoint.

I'd suggest simply using query parameters `lng` and `ns` to avoid having to parse the path. We might eventually support multi-loading, and the query parameters would fit better with that (although that could be difficult if we need to proxy some requests).

https://github.com/i18next/i18next-http-backend#backend-options

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -26,6 +26,79 @@ func NewPluginsHandler(client *http.Client, token string, pluginsEndpointMap map
+		klog.Errorf(redirectTo)
+	}
+
+	resp, err := p.Client.Get(redirectTo)
```
Is this actually making an HTTP request for static files? Can we serve static files directly instead using `http.ServeFile`?

jhadvig replied:
we should

spadgett reviewed (commented):
Thanks, @jhadvig. We'll want to update the frontend code as well to use the new endpoint.

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,15 +17,42 @@ import (
+	// In case of the dynamic plugin the localization namespace should contain name of the plugin prefixed with 'plugin:' prefix
+	// No file extension is needed since for the dynamic plugins we will fetch 'default.json' from the plugin pod.
+	// eg. 'plugin:helm-plugin'
+	pluginLocale := query.Get("ns")
```
This isn't a locale. I'd stick to `namespace` as the variable name.

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,15 +17,42 @@ import (
+	// In case of static plugin the localization namespace should contain name of the json file that contains the localization strings,
+	// since various static plugins have different names for the localization file.
+	// eg. namespace query for the local-storage-operator-plugin locales - '&ns=lso-plugin.json'
+	// In case of the dynamic plugin the localization namespace should contain name of the plugin prefixed with 'plugin:' prefix
+	// No file extension is needed since for the dynamic plugins we will fetch 'default.json' from the plugin pod.
+	// eg. 'plugin:helm-plugin'
```
Can you open a PR to update the dynamic plugin enhancement with this proposal? Then we can get agreement from stakeholders.

cc @christianvogt @vojtechszocs

jhadvig replied:
@spadgett https://github.com/openshift/enhancements/pull/815

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,15 +17,42 @@ import (
+	}
+	// in case of dynamic-plugin we need to trim the "plugin:" prefix.
+	pluginName := strings.TrimPrefix(pluginLocale, "plugin:")
+	http.Redirect(w, r, path.Join(r.Host, "api/plugins", pluginName, "locales", lang, "default.json"), http.StatusSeeOther)
```
We should confirm that i18next-http-backend handles redirects. It might be better just to proxy the content here if it's not too hard.

jhadvig replied:
yeah, proxying makes more sense ... updated

jhadvig commented:
@spadgett comments addressed.

jhadvig commented:
@spadgett updated the PR per updates in the enhancement

jhadvig commented:
/test e2e-gcp-console

jhadvig commented:
@spadgett I've addted locales to the dynamic-demo-plugin and updated the two Dockerfiles we have for it.

spadgett commented:
@vojtechszocs @christianvogt FYI

vojtechszocs commented on frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "Subject": "Subject"
```
Nit: please use 2 spaces for indentation (also in other JSON files).

vojtechszocs commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,73 @@ import (
+
+	query := r.URL.Query()
+	lang := query.Get("lng")
+	// In case of static plugin the localization namespace should contain name of the json file that contains the localization strings,
```
```suggestion
	// In case of console app or static plugins, the localization namespace should contain name of the json file that contains the localized strings,
```

vojtechszocs commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,73 @@ import (
+	query := r.URL.Query()
+	lang := query.Get("lng")
+	// In case of static plugin the localization namespace should contain name of the json file that contains the localization strings,
+	// since various static plugins have different names for the localization file.
```
```suggestion
	// since various console packages have different names for the localization file.
```

vojtechszocs commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,73 @@ import (
+	// eg. namespace query for the local-storage-operator-plugin locales - '&ns=lso-plugin.json'
+	// In case of the dynamic plugin the localization namespace should contain name of the plugin prefixed with 'plugin__' prefix.
+	// No file extension is needed since for the dynamic plugins we will fetch file from the plugin pod, based on the namespace.
+	// eg. 'plugin__helm-plugin' will fetch `helm-plugin.json`
```
```suggestion
	// eg. 'plugin__helm-plugin' will fetch `/locales/{lang}/helm-plugin.json`
```

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,73 @@ import (
 }
 
-func (p *PluginsHandler) HandlePlugins(w http.ResponseWriter, r *http.Request) {
+func (p *PluginsHandler) HandlePluginLocales(w http.ResponseWriter, r *http.Request) {
```
Naming nit: I'd call this `HandleLocales` since it handles both plugin locales and console app locales.

spadgett commented on frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "Subject": "Subject"
```
Is this used anywhere in the demo plugin?

I would only add the JSON files if we're actually using it somewhere in the demo plugin to give a real example.

jhadvig replied:
no its not, in that case I can just put there and empty json

vojtechszocs replied:
We can have a follow-up PR that illustrates localization usage in dynamic demo plugin.

There should be two examples:
- localizing strings in `console-extensions.json`
- localizing strings in a plain function
- localizing strings in a React component

Once this PR is merged and the conventions are settled, I can write that PR.

vojtechszocs commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,73 @@ import (
+		http.ServeFile(w, r, path.Join(p.PublicDir, "locales", lang, namespace))
+		return
+	}
+	// in case of dynamic-plugin we need to trim the "plugin__" prefix.
+	pluginName := strings.TrimPrefix(namespace, "plugin__")
```
On the plus side, this would prevent e.g. `plugin__kubevirt-plugin.json` file names :smiley: 

Having i18n namespace match the file name sounds pretty straight-forward, sounds OK to me.

jhadvig replied:
@spadgett I trim the `plugin__` prefix to get the plugin service endpoint form the map of all the active plugins, where we are mapping the plugin name to its service endpoint. Based on the conversation I dont think ConsolePlugin resource name will contain the prefix, thats why Im trimming it to get the real name.

jhadvig commented:
@spadgett @vojtechszocs I've rebased the PR and addressed comments. Also remove the i18n changes in the dynamic plugin since we will address those as follow up

PTAL

spadgett commented:
Test failure looks like it might be related.

> Error: Failed: Missing i18n key "YAML" in namespace "public" and language "en."

spadgett commented:
browser.log has a large amount of missing key errors:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/9196/pull-ci-openshift-console-master-e2e-gcp-console/1415975376487714816/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/browser.log

spadgett commented:
Looking deeper, I see the files are failing to load.

```
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=ceph-storage-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=console-shared - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=devconsole - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=container-security - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=console-app - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=git-service - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=gitops-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=helm-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=insights-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=kubevirt-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=knative-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=lso-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=metal3-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=notification-drawer - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=olm - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=public - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=pipelines-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=rhoas-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=topology - Failed to load resource: the server responded with a status of 404 (Not Found)
```

jhadvig commented:
Hmm wondering if thats related to the PR since there are also other errors

```
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/console/knative-event-sources - Failed to load resource: the server responded with a status of 401 (Unauthorized)
[WARNING] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js 0:380641 "Error fetching CRDs for dynamic event sources" Error: Unauthorized
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123823
    at Generator.next (<anonymous>)
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123209
    at new Promise (<anonymous>)
    at s (https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:122957)
    at u (https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123309)
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:124936
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/console/knative-channels - Failed to load resource: the server responded with a status of 401 (Unauthorized)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/vendors~main-chunk-d7eaff6a7eb98f5b246e.min.js 87 WebSocket connection to 'wss://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/graphql' failed: HTTP Authentication failed; no valid credentials available
```

spadgett commented:
@jhadvig Those are expected on initial console load since you're not logged in yet.

jhadvig commented:
/test e2e-gcp-console

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,74 @@ import (
 }
 
-func (p *PluginsHandler) HandlePlugins(w http.ResponseWriter, r *http.Request) {
+func (p *PluginsHandler) HandleLocales(w http.ResponseWriter, r *http.Request) {
```
These are resources, not locales.

```suggestion
func (p *PluginsHandler) HandleI18nResources(w http.ResponseWriter, r *http.Request) {
```

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,74 @@ import (
+
+	query := r.URL.Query()
+	lang := query.Get("lng")
+	// In case of console app or static plugins, the localization namespace should contain name of the json file that contains the localized strings,
+	// since various console packages have different names for the localization file.
+	// eg. namespace query for the local-storage-operator-plugin locales - '&ns=lso-plugin'
+	// In case of the dynamic plugin the localization namespace should contain name of the plugin prefixed with 'plugin__' prefix.
+	// No file extension is needed since for the dynamic plugins we will fetch file from the plugin pod, based on the namespace.
+	// eg. 'plugin__helm' will fetch `/locales/{lang}/plugin__helm.json`
```
I think this is the important bit, so I'd trim the comment to just this:

```suggestion
	// In case of the dynamic plugins, the namespace should contain name of the plugin prefixed with 'plugin__' prefix.
	// eg. 'plugin__helm' will fetch `locales/{lang}/plugin__helm.json` from the plugin service
```

spadgett commented on pkg/plugins/handlers.go:
```diff
@@ -17,32 +17,74 @@ import (
+	// In case of the dynamic plugin the localization namespace should contain name of the plugin prefixed with 'plugin__' prefix.
+	// No file extension is needed since for the dynamic plugins we will fetch file from the plugin pod, based on the namespace.
+	// eg. 'plugin__helm' will fetch `/locales/{lang}/plugin__helm.json`
+	namespace := query.Get("ns")
```
We should return 400 bad request if `lng` or `ns` are missing.

spadgett reviewed (approved):
/lgtm
/hold for approvals
/assign @yapei @ahardin-rh @sferich888

ahardin-rh commented:
/label docs-approved

sferich888 commented:
/label px-approved
/unassign @sferich888

yapei commented:
tested the changes and it is working well
/label qe-approved

jhadvig commented:
/hold cancel
